packages/flutter/lib/src/rendering/selection.dart

```diff
@@ -375,6 +399,114 @@ class SelectionEdgeUpdateEvent extends SelectionEvent {
   final Offset globalPosition;
 }
 
+/// Extends the start or end of the selection by a given [TextGranularity].
+///
+/// To handle this event, move the associated selection edge, as dictated by
+/// [isEnd], according to the [granularity].
+class GranularlyExtendSelectionEvent extends SelectionEvent {
+  /// Creates a [GranularlyExtendSelectionEvent].
+  ///
+  /// All parameters are required and must not be null.
+  const GranularlyExtendSelectionEvent({
+    required this.forward,
+    required this.isEnd,
+    required this.granularity,
+  }) : super._(SelectionEventType.granularlyExtendSelection);
+
+  /// Whether to extend the selection forward.
+  final bool forward;
+
+  /// Whether this event is updating the end selection edge.
+  final bool isEnd;
+
+  /// The granularity for which the selection extend.
+  final TextGranularity granularity;
+}
+
+/// The directional to extend a selection.
+///
+/// The [DirectionallyExtendSelectionEvent] uses this enum to describe how
+/// [Selectable] should extend their selection.
+enum SelectionExtendDirection {
+  /// Move one edge of the selection vertically to the previous adjacent line.
+  ///
+  /// See [DirectionallyExtendSelectionEvent.dx] on how to
+  /// calculate the horizontal offset.
```

Comment:
Nit: If you think it would be helpful, maybe clarify that this means the previous word-wrapped line, and not the previous hard linebreak?  Same for nextLine.